Replace govuk-lint with rubocop-govuk

laurentqro opened the pull request:
Replace deprecated govuk-lint with rubocop-govuk https://github.com/alphagov/rubocop-govuk, and fixes the shiny new linting errors.

Commit:
Replace deprecated govuk-lint with rubocop-govuk

Commit:
Replace single quoted strings with double-quoted ones

Fixes linting error: Prefer double-quoted strings unless you
need single quotes to avoid extra backslashes for escaping.
(https://rubystyle.guide#consistent-string-literals)

benthorner reviewed (commented):
What did I do to deserve this? 😉 

It looks like there are a couple of extra changes in here, so worth asking if they're intentional:

   - Previously we only linted specific directories, but now we're linting everything. I prefer this myself, because it makes it easier to auto-correct errors after running `rake lint`. Not that I write bad code, of course. Note that, unless it's [overridden in the Jenkinsfile](https://github.com/alphagov/content-publisher/blob/master/Jenkinsfile#L17), we'll only lint those old dirs on CI.

   - Previously we may have been running the Rails cops as well. I'm not sure. But I wonder if it's worth turning them on, as suggested [in the install instructions](https://github.com/alphagov/rubocop-govuk#installation). Depends on the number of errors to fix!

Commit:
Turn on Rails cops

Commit:
Use named HTTP status code

Commit:
Use Rails.root.join

Commit:
Use Time.zone.parse instead of Time.parse

Commit:
Replace html_safe with safe_join and content_tag

More info: https://www.rubydoc.info/gems/rubocop/0.41.2/RuboCop/Cop/Rails/OutputSafety

Commit:
Set :environment task as a dependency to all rake task

Commit:
Avoid using an ActiveRecord method

Commit:
Use #update instead of #update_attributes

Commit:
Remove unused methods in ApplicationHelper

This was moved to the StateHelper

Commit:
Move content_tag to the view

Commit:
Do not use instance variables in helpers

Relying on instance variables makes it difficult to re-use helper methods.
If it seems awkward to explicitly pass in each dependent variable,
consider moving the behaviour elsewhere, for example to a model, decorator or presenter.

https://www.rubydoc.info/gems/rubocop-rails/RuboCop/Cop/Rails/HelperInstanceVariable

benthorner commented on app/helpers/publishing_helper.rb:
```diff
@@ -1,30 +1,28 @@
       true
     rescue GdsApi::HTTPErrorResponse => e
-      error_response_message(:base, e.message) if @publishable
+      error_response_message(document, :base, e.message)
```
Are you sure we don't need the `if`?

laurentqro replied:
The value of`@publishable` (a Document instance) is currently assigned in `set_errors`, and there only.

`handle_remote_error` needs this `if @publishable` because `error_response_message` calls the #errors method on it.

If we pass the Document instance to both `handle_remote_error` and `error_response_message`, I don't think we need presence check anymore.

What do you think?

benthorner replied:
Sounds OK to me - worth a check 👍

benthorner commented on app/presenters/actions_presenter.rb:
```diff
@@ -18,21 +19,26 @@ def publish_button_visible?
 
   def publish_text
     if state == "published"
```
Much better! I think we can simplify this even further by removing the `text` variable, since the `if` will implicitly return its last expression.

laurentqro replied:
Good point!

benthorner commented on app/controllers/documents_controller.rb:
```diff
@@ -104,23 +106,13 @@ def check_authorisation
-      list_items.join.html_safe
+    heading = content_tag(:h4, "Please fix the following errors")
+    errors = content_tag(:ul, class: "list-unstyled remove-bottom-margin") do
+      safe_join(@document.errors.full_messages.map { |message| content_tag(:li, message.html_safe) })
```
I think we can do `raw` instead of `html_safe` here, since I'm guessing `message` is just a string.

benthorner commented on app/controllers/attachments_controller.rb:
```diff
@@ -31,7 +31,7 @@ def edit
     document = fetch_document
     attachment = document.attachments.find(attachment_content_id)
-    attachment.update_attributes(attachment_params)
+    attachment.update_properties(attachment_params)
```
Good fix - shame we didn't use `ActiveModel` in the first place 😠.

benthorner reviewed (commented):
Thanks for fixing all the errors @laurentqro! And the commits made it easy to review, too 😃. With the `html_safe` errors, I've commented on one specifically already; with the Govspeak one, I'd suggest [ignoring the error](https://github.com/alphagov/content-publisher/blob/master/app/controllers/govspeak_preview_controller.rb#L10), which should be fine since Govspeak should be sanitising stuff by default.

Commit:
Remove unnecessary variable assignment

The `text` variable is not needed, since the
`if` implicitly returns its last expression.

Commit:
Ignore Rails/OutputSafety error

It should be fine to do this, since Govspeak should be sanitising stuff by default.

Commit:
Ignore Rails/OutputSafety error because messages are only strings

barrucadu reviewed (approved):
LGTM
